packages/clerk-js/src/utils/passkeys.ts

```diff
@@ -0,0 +1,191 @@
+    const res = credential.response as AuthenticatorAttestationResponse;
+
+    return { publicKeyCredential: { ...credential, response: res }, error: null };
+  } catch (e) {
+    return { error: handlePublicKeyCreateError(e), publicKeyCredential: null };
+  }
+}
+
+/**
+ * Map webauthn errors from `navigator.credentials.create()` to Clerk-js errors
+ * @param error
+ */
+function handlePublicKeyCreateError(error: Error): ClerkWebAuthnError | ClerkRuntimeError | Error {
+  if (error.name === 'InvalidStateError') {
+    return new ClerkWebAuthnError(error.message, { code: 'passkey_exists' });
+  } else if (error.name === 'NotAllowedError') {
+    return new ClerkWebAuthnError(error.message, { code: 'passkey_registration_cancelled' });
+  }
+  return error;
+}
+
+function convertJSONToPublicKeyCreateOptions(jsonPublicKey: PublicKeyCredentialCreationOptionsJSON) {
+  const userIdBuffer = base64UrlToBuffer(jsonPublicKey.user.id);
+  const challengeBuffer = base64UrlToBuffer(jsonPublicKey.challenge);
+
+  const excludeCredentialsWithBuffer = (jsonPublicKey.excludeCredentials || []).map(cred => ({
+    ...cred,
+    id: base64UrlToBuffer(cred.id),
+  }));
+
+  return {
+    ...jsonPublicKey,
+    excludeCredentials: excludeCredentialsWithBuffer,
+    challenge: challengeBuffer,
+    user: {
+      ...jsonPublicKey.user,
+      id: userIdBuffer,
+    },
+  } as PublicKeyCredentialCreationOptionsWithoutExtensions;
+}
+
+function serializePublicKeyCredential(publicKeyCredential: PublicKeyCredentialWithAuthenticatorAttestationResponse) {
+  const response = publicKeyCredential.response;
+  return {
+    type: publicKeyCredential.type,
+    id: publicKeyCredential.id,
+    rawId: bufferToBase64Url(publicKeyCredential.rawId),
+    authenticatorAttachment: publicKeyCredential.authenticatorAttachment,
+    response: {
+      clientDataJSON: bufferToBase64Url(response.clientDataJSON),
+      attestationObject: bufferToBase64Url(response.attestationObject),
+      transports: response.getTransports(),
+    },
+  };
+}
+
+const bufferToBase64Url = Base64Converter.encode.bind(Base64Converter);
+const base64UrlToBuffer = Base64Converter.decode.bind(Base64Converter);
+
+/**
+ * Custom error class for representing Clerk runtime errors.
+ *
+ * @class ClerkRuntimeError
+ * @example
+ *   throw new ClerkRuntimeError('An error occurred', { code: 'password_invalid' });
+ */
+export class ClerkWebAuthnError extends ClerkRuntimeError {
+  /**
+   * A unique code identifying the error, used for localization
+   *
+   * @type {string}
+   * @memberof ClerkRuntimeError
+   */
+  code: ClerkWebAuthnErrorCode;
+
+  constructor(message: string, { code }: { code: ClerkWebAuthnErrorCode }) {
+    super(message, { code });
+
+    Object.setPrototypeOf(this, ClerkWebAuthnError.prototype);
+
+    this.code = code;
+    this.message = message;
+    this.clerkRuntimeError = true;
+  }
+}
```

Comment:
This could be removed, my goal was to improve typesafety for error codes regarding passkeys